packages/vite/src/node/plugins/html.ts

```diff
@@ -1265,6 +1265,45 @@ export function resolveHtmlTransforms(
   return [preHooks, normalHooks, postHooks]
 }
 
+// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/head#see_also
+const HeadElement: Record<string, boolean> = [
+  'title',
+  'base',
+  'link',
+  'style',
+  'meta',
+  'script',
+  'noscript',
+  'template',
+].reduce(
+  (res, tag) => ({
+    ...res,
+    [tag]: true,
+  }),
+  {},
+)
+
+function headTagInsertCheck(
+  tags: HtmlTagDescriptor[],
+  ctx: IndexHtmlTransformContext,
+) {
+  if (!tags.length) return
+  const { logger } = ctx.server?.config || {}
+  const message = tags.reduce<string[]>((res, tagDescriptor) => {
+    if (!HeadElement[tagDescriptor.tag]) res.push(`<${tagDescriptor.tag}>`)
+    return res
+  }, [])
+  if (message.length) {
+    logger?.warn(
+      colors.yellow(
+        colors.bold(
+          `[${message.join(',')}] can not be used inside the <head> Element, please check the 'injectTo' value`,
```

Comment:
```suggestion
  const disallowedTags = tags.filter(tagDescriptor => !elementsAllowedInHead.has(tagDescriptor.tag))
  if (disallowedTags.length) {
    const dedupedTags = unique(disallowedTags.map(tag => `<${tagDescriptor.tag}>`))
    logger?.warn(
      colors.yellow(
        colors.bold(
          `[${dedupedTags.join(',')}] can not be used inside the <head> Element, please check the 'injectTo' value`,
```
Let's avoid `.reduce` for readability. Also added `unique` to dedupe tags.

Reply:
Done